cpp/src/hash/hashing.cu

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/detail/hashing.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/nvtx_utils.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/detail/scatter.hpp>
+
+#include <thrust/tabulate.h>
+
+namespace cudf {
+
+namespace {
+
+/** 
+ * @brief  Functor to map a hash value to a particular 'bin' or partition number
+ * that uses the modulo operation.
+ */
+template <typename hash_value_t>
+struct modulo_partitioner
+{
+  modulo_partitioner(size_type num_partitions) : divisor{num_partitions} {}
+
+  __host__ __device__
+  size_type operator()(hash_value_t hash_value) const {
+    return hash_value % divisor;
+  }
+
+  const size_type divisor;
+};
+
+template <typename T>
+bool is_power_two(T number) {
+  return (0 == (number & (number - 1)));
+}
+
+/** 
+ * @brief  Functor to map a hash value to a particular 'bin' or partition number
+ * that uses bitshifts. Only works when num_partitions is a power of 2.
+ *
+ * For n % d, if d is a power of two, then it can be computed more efficiently via 
+ * a single bitwise AND as:
+ * n & (d - 1)
+ */
+template <typename hash_value_t>
+struct bitwise_partitioner
+{
+  bitwise_partitioner(size_type num_partitions) : divisor{(num_partitions - 1)} {
+    assert(is_power_two(num_partitions));
+  }
+
+  __host__ __device__
+  size_type operator()(hash_value_t hash_value) const {
+    return hash_value & (divisor);
```

Comment:
Should this be?
```suggestion
    return hash_value & (divisor-1);
```

Reply:
Divisor is already of the type 2^n - 1 so I think this is fine. Perhaps the documentation can use explicit variable names to clear this out.

Reply:
Ok, I made this more clear with a comment and change of variable name: `return hash_value & mask; // hash_value & (num_partitions - 1)`